File: tests/unit/test_trit.c
/**
 * @file test_trit.c
 * @brief Unit tests for Trit (ternary value) implementation
 */

#include "../include/kernel/trit.h"
#include <stdio.h>
#include <assert.h>
#include <string.h>

// Test counter
static int tests_run = 0;
static int tests_passed = 0;

// Test assertion macro
#define TEST_ASSERT(condition, message) \
    do { \
        tests_run++; \
        if (condition) { \
            printf("  ✓ %s\n", message); \
            tests_passed++; \
        } else { \
            printf("  ✗ %s\n", message); \
        } \
    } while(0)

// Test Trit creation
void test_trit_creation() {
    printf("Testing Trit creation...\n");
    
    trit_t t1 = trit_create(TERNARY_POSITIVE);
    TEST_ASSERT(trit_get_value(t1) == TERNARY_POSITIVE, "Create positive trit");
    TEST_ASSERT(trit_is_valid(t1), "Positive trit is valid");
    
    trit_t t2 = trit_create(TERNARY_NEUTRAL);
    TEST_ASSERT(trit_get_value(t2) == TERNARY_NEUTRAL, "Create neutral trit");
    TEST_ASSERT(trit_is_valid(t2), "Neutral trit is valid");
    
    trit_t t3 = trit_create(TERNARY_NEGATIVE);
    TEST_ASSERT(trit_get_value(t3) == TERNARY_NEGATIVE, "Create negative trit");
    TEST_ASSERT(trit_is_valid(t3), "Negative trit is valid");
}

// Test Trit logic operations
void test_trit_logic() {
    printf("Testing Trit logic operations...\n");
    
    trit_t t_pos = trit_create(TERNARY_POSITIVE);
    trit_t t_neu = trit_create(TERNARY_NEUTRAL);
    trit_t t_neg = trit_create(TERNARY_NEGATIVE);
    
    // Test AND
    trit_t result = trit_and(t_pos, t_pos);
    TEST_ASSERT(trit_get_value(result) == TERNARY_POSITIVE, "AND: 1 AND 1 = 1");
    
    result = trit_and(t_neg, t_neg);
    TEST_ASSERT(trit_get_value(result) == TERNARY_NEGATIVE, "AND: -1 AND -1 = -1");
    
    result = trit_and(t_pos, t_neg);
    TEST_ASSERT(trit_get_value(result) == TERNARY_NEGATIVE, "AND: 1 AND -1 = -1");
    
    // Test OR
    result = trit_or(t_pos, t_pos);
    TEST_ASSERT(trit_get_value(result) == TERNARY_POSITIVE, "OR: 1 OR 1 = 1");
    
    result = trit_or(t_pos, t_neg);
    TEST_ASSERT(trit_get_value(result) == TERNARY_POSITIVE, "OR: 1 OR -1 = 1");
    
    // Test NOT
    result = trit_not(t_neg);
    TEST_ASSERT(trit_get_value(result) == TERNARY_POSITIVE, "NOT: ~(-1) = 1");

    result = trit_not(t_pos);
    TEST_ASSERT(trit_get_value(result) == TERNARY_NEGATIVE, "NOT: ~(1) = -1");
    
    result = trit_not(t_neu);
    TEST_ASSERT(trit_get_value(result) == TERNARY_NEUTRAL, "NOT: ~(0) = 0");
    
    // Test XOR
    result = trit_xor(t_pos, t_pos);
    TEST_ASSERT(trit_get_value(result) == TERNARY_NEUTRAL, "XOR: 1 XOR 1 = 0");
    
    result = trit_xor(t_pos, t_neg);
    TEST_ASSERT(trit_get_value(result) == TERNARY_POSITIVE, "XOR: 1 XOR -1 = 1");
}

// Test Trit arithmetic operations
void test_trit_arithmetic() {
    printf("Testing Trit arithmetic operations...\n");
    
    trit_t t_pos = trit_create(TERNARY_POSITIVE);
    trit_t t_neu = trit_create(TERNARY_NEUTRAL);
    trit_t t_neg = trit_create(TERNARY_NEGATIVE);
    
    // Test ADD
    trit_t result = trit_add(t_pos, t_neu);
    TEST_ASSERT(trit_get_value(result) == TERNARY_POSITIVE, "ADD: 1 + 0 = 1");
    
    result = trit_add(t_pos, t_neg);
    TEST_ASSERT(trit_get_value(result) == TERNARY_NEUTRAL, "ADD: 1 + (-1) = 0");
    
    result = trit_add(t_neg, t_neg);
    TEST_ASSERT(trit_get_value(result) == TERNARY_NEGATIVE, "ADD: -1 + (-1) = -1");
    
    // Test SUB
    result = trit_subtract(t_pos, t_neu);
    TEST_ASSERT(trit_get_value(result) == TERNARY_POSITIVE, "SUB: 1 - 0 = 1");
    
    result = trit_subtract(t_pos, t_pos);
    TEST_ASSERT(trit_get_value(result) == TERNARY_NEUTRAL, "SUB: 1 - 1 = 0");
    
    // Test MUL
    result = trit_multiply(t_pos, t_pos);
    TEST_ASSERT(trit_get_value(result) == TERNARY_POSITIVE, "MUL: 1 * 1 = 1");
    
    result = trit_multiply(t_neg, t_pos);
    TEST_ASSERT(trit_get_value(result) == TERNARY_NEGATIVE, "MUL: -1 * 1 = -1");
    
    result = trit_multiply(t_pos, t_neu);
    TEST_ASSERT(trit_get_value(result) == TERNARY_NEUTRAL, "MUL: 1 * 0 = 0");
}

int main() {
    printf("========================================\n");
    printf("TEROS Trit Unit Tests\n");
    printf("========================================\n\n");
    
    test_trit_creation();
    printf("\n");
    
    test_trit_logic();
    printf("\n");
    
    test_trit_arithmetic();
    printf("\n");
    
    printf("========================================\n");
    printf("Tests run: %d\n", tests_run);
    printf("Tests passed: %d\n", tests_passed);
    printf("Tests failed: %d\n", tests_run - tests_passed);
    printf("========================================\n");
    
    return (tests_run == tests_passed) ? 0 : 1;
}

File: tests/unit/test_trit.py
"""
Unit tests for Trit (ternary value) operations.

Tests cover:
- Basic trit operations (AND, OR, NOT, XOR)
- Arithmetic operations (ADD, SUB, MUL, DIV)
- Comparison and validation
- Edge cases and error handling
"""

import pytest
import sys
from pathlib import Path

# Add project root to path
sys.path.insert(0, str(Path(__file__).parent.parent.parent))

# Try importing Python implementation first
try:
    from src.lib.teros.core.trit import Trit
    PYTHON_AVAILABLE = True
except ImportError:
    PYTHON_AVAILABLE = False

# TODO: Import C implementation when available
# from src.kernel.trit import trit_t, trit_create


@pytest.mark.unit
class TestTritCreation:
    """Test trit creation and initialization."""
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_create_positive(self):
        """Test creating positive trit (+1)."""
        trit = Trit(1)
        assert trit.value == 1
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_create_negative(self):
        """Test creating negative trit (-1)."""
        trit = Trit(-1)
        assert trit.value == -1
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_create_neutral(self):
        """Test creating neutral trit (0)."""
        trit = Trit(0)
        assert trit.value == 0
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_invalid_value(self):
        """Test that invalid values raise exception."""
        with pytest.raises(ValueError):
            Trit(2)
        with pytest.raises(ValueError):
            Trit(-2)


@pytest.mark.unit
class TestTritLogic:
    """Test ternary logic operations."""
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_and(self):
        """Test ternary AND operation."""
        assert (Trit(-1) & Trit(-1)).value == -1
        assert (Trit(-1) & Trit(0)).value == 0
        assert (Trit(-1) & Trit(1)).value == -1
        assert (Trit(0) & Trit(-1)).value == 0
        assert (Trit(0) & Trit(0)).value == 0
        assert (Trit(0) & Trit(1)).value == 0
        assert (Trit(1) & Trit(-1)).value == -1
        assert (Trit(1) & Trit(0)).value == 0
        assert (Trit(1) & Trit(1)).value == 1
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_or(self):
        """Test ternary OR operation."""
        assert (Trit(-1) | Trit(-1)).value == -1
        assert (Trit(-1) | Trit(0)).value == -1
        assert (Trit(-1) | Trit(1)).value == 1
        assert (Trit(0) | Trit(-1)).value == -1
        assert (Trit(0) | Trit(0)).value == 0
        assert (Trit(0) | Trit(1)).value == 1
        assert (Trit(1) | Trit(-1)).value == 1
        assert (Trit(1) | Trit(0)).value == 1
        assert (Trit(1) | Trit(1)).value == 1
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_not(self):
        """Test ternary NOT operation."""
        assert (~Trit(-1)).value == 1
        assert (~Trit(0)).value == 0
        assert (~Trit(1)).value == -1
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_xor(self):
        """Test ternary XOR operation."""
        assert (Trit(-1) ^ Trit(-1)).value == 0
        assert (Trit(-1) ^ Trit(0)).value == -1
        assert (Trit(-1) ^ Trit(1)).value == 1
        assert (Trit(0) ^ Trit(-1)).value == -1
        assert (Trit(0) ^ Trit(0)).value == 0
        assert (Trit(0) ^ Trit(1)).value == 1
        assert (Trit(1) ^ Trit(-1)).value == 1
        assert (Trit(1) ^ Trit(0)).value == 1
        assert (Trit(1) ^ Trit(1)).value == 0


@pytest.mark.unit
class TestTritArithmetic:
    """Test ternary arithmetic operations."""
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_add(self):
        """Test ternary addition."""
        assert (Trit(-1) + Trit(-1)).value == -1
        assert (Trit(-1) + Trit(0)).value == -1
        assert (Trit(-1) + Trit(1)).value == 0
        assert (Trit(0) + Trit(-1)).value == -1
        assert (Trit(0) + Trit(0)).value == 0
        assert (Trit(0) + Trit(1)).value == 1
        assert (Trit(1) + Trit(-1)).value == 0
        assert (Trit(1) + Trit(0)).value == 1
        assert (Trit(1) + Trit(1)).value == 1
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_subtract(self):
        """Test ternary subtraction."""
        assert (Trit(-1) - Trit(-1)).value == 0
        assert (Trit(-1) - Trit(0)).value == -1
        assert (Trit(-1) - Trit(1)).value == -1
        assert (Trit(0) - Trit(-1)).value == 1
        assert (Trit(0) - Trit(0)).value == 0
        assert (Trit(0) - Trit(1)).value == -1
        assert (Trit(1) - Trit(-1)).value == 1
        assert (Trit(1) - Trit(0)).value == 1
        assert (Trit(1) - Trit(1)).value == 0
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_multiply(self):
        """Test ternary multiplication."""
        assert (Trit(-1) * Trit(-1)).value == 1
        assert (Trit(-1) * Trit(0)).value == 0
        assert (Trit(-1) * Trit(1)).value == -1
        assert (Trit(0) * Trit(-1)).value == 0
        assert (Trit(0) * Trit(0)).value == 0
        assert (Trit(0) * Trit(1)).value == 0
        assert (Trit(1) * Trit(-1)).value == -1
        assert (Trit(1) * Trit(0)).value == 0
        assert (Trit(1) * Trit(1)).value == 1


@pytest.mark.unit
class TestTritConversion:
    """Test trit conversions."""
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_to_int(self):
        """Test conversion to integer."""
        assert int(Trit(-1)) == -1
        assert int(Trit(0)) == 0
        assert int(Trit(1)) == 1
    
    @pytest.mark.skipif(not PYTHON_AVAILABLE, reason="Python trit implementation not available")
    def test_to_bool(self):
        """Test conversion to boolean."""
        assert bool(Trit(-1)) == False
        assert bool(Trit(0)) == False
        assert bool(Trit(1)) == True


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

